File: python/day02-control-flow/solutions.py
"""Day 2 reference solutions."""


def classify_number(n):
    if n < 0:
        return "negative"
    elif n == 0:
        return "zero"
    else:
        return "positive"


def fizzbuzz(n):
    result = []
    for i in range(1, n + 1):
        if i % 15 == 0:
            result.append("FizzBuzz")
        elif i % 3 == 0:
            result.append("Fizz")
        elif i % 5 == 0:
            result.append("Buzz")
        else:
            result.append(str(i))
    return result


def first_matching_index(items, predicate):
    for i, item in enumerate(items):
        if predicate(item):
            return i
    else:
        return -1


def squares_of_evens(numbers):
    return [n * n for n in numbers if n % 2 == 0]


def word_lengths(words):
    return {word: len(word) for word in words}


def describe_shape(shape):
    match shape:
        case ("circle", radius):
            return f"circle with radius {radius}"
        case ("rectangle", w, h):
            return f"rectangle {w}x{h}"
        case ("square", side):
            return f"square with side {side}"
        case _:
            return "unknown shape"

File: typescript-javascript/day02-control-flow/solutions.js
// Day 2 reference solutions.

function classifyNumber(n) {
    if (n < 0) {
        return "negative";
    } else if (n === 0) {
        return "zero";
    } else {
        return "positive";
    }
}

function fizzbuzz(n) {
    const result = [];
    for (let i = 1; i <= n; i++) {
        if (i % 15 === 0) {
            result.push("FizzBuzz");
        } else if (i % 3 === 0) {
            result.push("Fizz");
        } else if (i % 5 === 0) {
            result.push("Buzz");
        } else {
            result.push(String(i));
        }
    }
    return result;
}

function sumUntilNegative(numbers) {
    let total = 0;
    for (const n of numbers) {
        if (n < 0) {
            break;
        }
        total += n;
    }
    return total;
}

function describeShape(shape) {
    switch (shape.type) {
        case "circle":
            return `circle with radius ${shape.radius}`;
        case "rectangle":
            return `rectangle ${shape.width}x${shape.height}`;
        case "square":
            return `square with side ${shape.side}`;
        default:
            return "unknown shape";
    }
}

function countVowels(word) {
    const vowels = "aeiou";
    let count = 0;
    for (const character of word) {
        if (vowels.includes(character)) {
            count++;
        }
    }
    return count;
}

module.exports = { classifyNumber, fizzbuzz, sumUntilNegative, describeShape, countVowels };
